src/scanpy/metrics/_gearys_c.py

```diff
@@ -109,22 +110,33 @@ def gearys_c(
         np.testing.assert_array_equal(pc_c, alt)
 
     """
-    if use_graph is None:
-        # Fix for anndata<0.7
-        if hasattr(adata, "obsp") and "connectivities" in adata.obsp:
-            g = adata.obsp["connectivities"]
-        elif "neighbors" in adata.uns:
-            g = adata.uns["neighbors"]["connectivities"]
-        else:
-            msg = "Must run neighbors first."
-            raise ValueError(msg)
-    else:
-        raise NotImplementedError()
+    g = _get_graph(adata, use_graph=use_graph)
     if vals is None:
         vals = _get_obs_rep(adata, use_raw=use_raw, layer=layer, obsm=obsm, obsp=obsp).T
     return gearys_c(g, vals)
 
 
+@gearys_c.register(sparse.csr_matrix)
+def _gearys_c(g: sparse.csr_matrix, vals: _Vals) -> NDArray:
+    return _GearysC(g, vals)()
+
+
+class _GearysC(_SparseMetric):
+    name = "Geary’s C"
+
+    def mtx(self, new_vals: NDArray | sparse.csr_matrix) -> NDArray:
+        g_parts = (self.g.data, self.g.indices, self.g.indptr)
+        if isinstance(new_vals, np.ndarray):
+            return _gearys_c_mtx(*g_parts, new_vals)
+        v_parts = (new_vals.data, new_vals.indices, new_vals.indptr)
+        return _gearys_c_mtx_csr(*g_parts, *v_parts, new_vals.shape)
```

Comment:
This and `vec` are basically identical between Moran's and Geary's unless I am missing something with the exception of the actual function i.e., they do this `g_parts` thing, have an instance check etc.  Maybe deduplicate?

Reply:
I’ll see what I can do. The issue is that numba functions necessarily need to have ugly signatures I think. I’ll see if I can improve that and if not, at least wrap them somehow.

Reply:
Hmm numba-scipy exists, but it seems pretty dead and has no support for csr_array/csc_array: https://github.com/numba/numba-scipy/issues/109

Maybe we should vendor its code into fast-array-utils and maintain it there?

Reply:
OK, @ilan-gold agreed, so I’ll leave that for now